### PR TITLE
[SPARK-12437][SQL] [WIP]Encapsulate the table and  column name in quotes

### PR DESCRIPTION
**Pull request by naveenminchu**
test this please

### CODE REVIEW

**Comment by hvanhovell** [comment]
ok to test

**Comment by hvanhovell** [comment]
@naveenminchu is there anyway sensible way to test this? If so, could you add a test?

**Comment by hvanhovell** [comment]
@naveenminchu I tried triggering a build, but it seems that I do not have sufficient rights for this (I do have them for my own PRs).

**Comment by rxin** [comment]
Different databases have different quoting syntax. You need to use the JDBCDialect's quoting function to quote the table/column names.

**Comment by naveenminchu** [comment]
@rxin Agree 100%

**Comment by JoshRosen** [comment]
Can you add a test for this in the MySQL docker integration tests suite (see the `docker-integration-tests` subproject)?

**Comment by JoshRosen** [comment]
Also, quick question: what if the name is qualified with a schema and also contains special characters? e.g. what if my user-supplied table name is something like `default."table with spaces in name"`? Will this handle that case properly? What about `"some database"."some-table-name"`? My concern is that a naive, unconditional quoting will mishandle these cases.

`spark-redshift` has some fancy logic to handle all of these corner-cases; see https://github.com/databricks/spark-redshift/pull/102 and pay attention to how the `TableName` wrapper works.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -311,7 +312,7 @@ final class DataFrameWriter private[sql](df: DataFrame) {
    * This is equivalent to:
    * {{{
    *   format("json").save(path)
-   * }}}
+   * }}}$table
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: what's this change for?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 2]
```diff
@@ -219,13 +219,21 @@ object JdbcUtils extends Logging {
...
       val nullable = if (field.nullable) "" else "NOT NULL"
       sb.append(s", $name $typ $nullable")
     }}
     if (sb.length < 2) "" else sb.substring(2)
   }
+  
+  /**
+   * Compute the table name string for this RDD.
+   */
+  def tableString(table: String, url: String): String = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: since this is only used once, can you just remove this function and inline it?

**Comment by rxin** [comment]
And +1 on adding a test.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 2 above)
Comment (naveenminchu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: @JoshRosen Are you suggesting to build similar logic as spark-redshift has some fancy logic to handle all of these corner-cases; that you highlighted databricks/spark-redshift#102 as to how the TableName wrapper works.

Please let me know ur thoughts

**Comment by rxin** [comment]
BTW if I understand this correctly it only quotes table names, but not column names? Are column names already quoted?

**Comment by naveenminchu** [comment]
@rxin Yes as u suggested I have used JdbcDialect

Will also look in to test part

**Comment by naveenminchu** [comment]
@rxin & @JoshRosen Please take a quick look of changes

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala [hunk 3]
```diff
@@ -150,4 +150,91 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationSuite {
     df2.write.jdbc(jdbcUrl, "datescopy", new Properties)
     df3.write.jdbc(jdbcUrl, "stringscopy", new Properties)
   }
+  
+  test("Write test with SaveMode set to overwrite") {
```
Comment (pjfanning) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala]: Would it be possible to add a table or tables to the dataPreparation function that have column names that are key words? For example:
create table `escaped names` (`key` BIGINT, `long description` TEXT);

Comment (naveenminchu) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala]: @pjfanning Added test as suggested 
For example:
create table `escaped names` (`key` BIGINT, `long description` TEXT);

**Comment by pjfanning** [comment]
Hi Naveen - your new tests look good to me. I do notice though that you probably need to resync with master due to conflicts (there is a 'This branch has conflicts that must be resolved' warning).
https://help.github.com/articles/syncing-a-fork/

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala [hunk 4]
```diff
@@ -41,6 +41,26 @@ private case object MySQLDialect extends JdbcDialect {
   override def quoteIdentifier(colName: String): String = {
     s"`$colName`"
   }
+  
+  override def parseTableName(tableName: String): String = {
+    val tableName1 = tableName.replace("\"", "").replace("\'", "")
+    if (tableName1.contains(".")) {
+      val tableNameList = tableName1.split('.')
+      tableNameList.foldLeft("") { (leftStr, rightStr) =>
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Woah, this is kinda dense and confusing. Can you please rewrite this in a simpler fashion and add comments? This is just my drive-by first impression.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Also, can you add unit tests for this function? _Not_ end-to-end tests, but a test which exercises this method in isolation, similar to the ones that I have in `spark-redshift`?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala [hunk 5]
```diff
@@ -41,6 +41,26 @@ private case object MySQLDialect extends JdbcDialect {
   override def quoteIdentifier(colName: String): String = {
     s"`$colName`"
   }
+  
+  override def parseTableName(tableName: String): String = {
+    val tableName1 = tableName.replace("\"", "").replace("\'", "")
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Why is it correct to just drop quotes here? What if the schema name also contains a dot? Don't you need to keep track of quotes during the string processing in order to figure out where the schema name ends and the table name begins?

Note that both the schema and table names could contain dots (AFAIK, if not please comment with a citation explaining why not), so simply grabbing the last component after a dot might not be correct.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala [hunk 6]
```diff
@@ -41,6 +41,26 @@ private case object MySQLDialect extends JdbcDialect {
...
+            leftStr + s"`$rightStr`"
+          } else {
+            leftStr + "." + s"`$rightStr`"
+          }
+        } else {
+          leftStr
+        }
+      }
+    } else {
+      s"`$tableName1`"
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Is this supposed to return quoted identifiers? Or should that be the responsibility of another function? Just want to make sure that we're not double-quoting.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala (hunk 5 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Also, `tableName1` is a bad name; if you were going to go with this approach, I'd call it something like `tableNameWithoutQuotes` (but I'm not sure that this unconditional quote-stripping is correct).

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 7]
```diff
@@ -90,6 +90,15 @@ abstract class JdbcDialect extends Serializable {
   }
 
   /**
+   * Parses a table name in case of containing special characters like . seperating database name
+   * followed by table name (eg "some database"."some-table-name") or 
+   * in case it contains characters that require quotes (e.g. space).
+   */
+  def parseTableName(tableName: String): String = {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala]: The method contract here is confusing to me: is this supposed to introduce an additional layer of quotes? I find it confusing to couple both the responsibilities of quoting identifiers and parsing table names from a schema-qualified name.

I'd prefer to have this return just the table name, unquoted, and leave it up to later components to escape it. If you _do_ think we should do quoting here, please explain that part of the API contract in the Scaladoc.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 8]
```diff
@@ -219,13 +219,21 @@ object JdbcUtils extends Logging {
...
+      val name = dialect.quoteIdentifier(field.name)
       val typ: String = getJdbcType(field.dataType, dialect).databaseTypeDefinition
       val nullable = if (field.nullable) "" else "NOT NULL"
       sb.append(s", $name $typ $nullable")
     }}
     if (sb.length < 2) "" else sb.substring(2)
   }
+  
+  /**
+   * Parse the table name string for this RDD.
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: This is confusing, since the variable `table` already sounds like it would be the table name.

Can you call it `schemaQualifiedTableName` or something more explicit?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala (hunk 6 above)
Comment (naveenminchu) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Yes its to return quoted identifiers

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (naveenminchu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Adjusted as suggested

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala [hunk 9]
```diff
@@ -40,6 +40,34 @@ private case object MySQLDialect extends JdbcDialect {
...
+   * in case it contains characters that require quotes (e.g. space).
+   */
+  override def schemaQualifiedTableName(tableName: String): String = {
+    //Removing quotes so that we can add them correctly.
+    val tableNameWithoutQuotes = tableName.replace("\"", "").replace("\'", "")
+    
+    //If block for addressing the case of . (eg "some database"."some-table-name")
+    if (tableNameWithoutQuotes.contains(".")) {
+      val tableNameList = tableNameWithoutQuotes.split('.')
+      tableNameList.foldLeft("") { (leftStr, rightStr) =>
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: I'm still not a fan of this `foldLeft`. Can you please write this in a more clear way?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Ping. Any updates here?

Comment (naveenminchu) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Will address it these week sometime
